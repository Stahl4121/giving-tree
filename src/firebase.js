//Reasoning for this file: https://stackoverflow.com/questions/48492047/where-do-i-initialize-firebase-app-in-react-application

// Firebase App (the core Firebase SDK) is always required and
// must be listed before other Firebase SDKs
import * as firebase from "firebase/app";

// Add the Firebase services that you want to use
import "firebase/auth";
import "firebase/firestore";

//Import everything, not recommended for production apps
//import firebase from "firebase";

const firebaseConfig = {
  apiKey: "AIzaSyAkeCK6w-qbvHzaOdEyg_PtpGQbElAv0_M",
  authDomain: "givingtree-cfs.firebaseapp.com",
  databaseURL: "https://givingtree-cfs.firebaseio.com",
  projectId: "givingtree-cfs",
  storageBucket: "givingtree-cfs.appspot.com",
  messagingSenderId: "709659064913",
  appId: "1:709659064913:web:d9decf22beb442f2e79670",
  measurementId: "G-KZ6PNG9K64"
};


firebase.initializeApp(firebaseConfig);

export default !firebase.apps.length ? firebase.initializeApp(firebaseConfig) : firebase.app();