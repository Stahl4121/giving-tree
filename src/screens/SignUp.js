import React, { Component } from 'react';
import { Link, withRouter } from 'react-router-dom';
import $ from 'jquery';

import firebase from '../firebase.js';
import * as helper from '../helpers/SignUpHelper.js';

import { withStyles } from '@material-ui/styles';
import Container from '@material-ui/core/Container';
import Avatar from '@material-ui/core/Avatar';
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Grid from "@material-ui/core/Grid";
import MUILink from "@material-ui/core/Link";
import Typography from "@material-ui/core/Typography";


const styles = theme => ({
    root: {
        marginTop: theme.spacing(8),
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
    },
    avatar: {
        margin: theme.spacing(1),
        backgroundColor: theme.palette.secondary.main,
    },
    form: {
        marginTop: theme.spacing(1),
    },
    submit: {
        margin: theme.spacing(3, 0, 2),
    },
    links: {
        textDecoration: 'none'
    },
    errorMsg: {
        color: 'red',
        marginTop: theme.spacing(1),
    }
});

function SignUp(props) {
    const { classes } = props;

    return (
        <Container component="main" maxWidth="xs">
            <div className={classes.root}>
                <Avatar className={classes.avatar}>
                    <LockOutlinedIcon />
                </Avatar>
                <Typography component="h1" variant="h5" className={classes.header}>
                    Sign Up
                    </Typography>
                <SignUpForm classes={classes} />
                <SignInForgotLink classes={classes} />
            </div>
        </Container>
    );
}

const INITIAL_STATE = {
    email: '',
    passwordOne: '',
    passwordTwo: '',
    errors: {
        email: '',
        passwordOne: '',
        passwordTwo: '',
    },
    submitError: ''
};

class SignUpFormBase extends Component {
    constructor(props) {
        super(props);
        this.state = { ...INITIAL_STATE };
    }

    onSubmit = event => {
        let errors = {
            email: helper.validateEmail(this.state.email),
            passwordOne: helper.validatePassword(this.state.passwordOne),
            passwordTwo: helper.confirmMatching(this.state.passwordOne, this.state.passwordTwo),
        }

        if (errors.email === '' && errors.passwordOne === '' && errors.passwordTwo === '') {
            const { email, passwordOne } = this.state;

            firebase.auth().createUserWithEmailAndPassword(email, passwordOne)
                .then((result) => {
                    //Do Misc Stuff for Donor Account
                    //Post to Server? Set firebase rules?

                    //-----------------------------------------
                    //--------MOVE INTO ACCOUNT REQUEST--------
                    //-----------------------------------------
                    // User is signed in. Get the ID token.
                    return result.user.getIdToken();
                })
                .then((idToken) => {
                    // Pass the ID token to the server.
                    $.post(
                        '/setCustomClaims',
                        {
                            idToken: idToken,
                        },
                        (data, status) => {
                            // This is not required. You could just wait until the token is expired
                            // and it proactively refreshes.
                            if (status === 'success' && data) {
                                const json = JSON.parse(data);
                                if (json && json.status === 'success') {
                                    // Force token refresh. The token claims will contain the additional claims.
                                    firebase.auth().currentUser.getIdToken(true);
                                }
                            }
                            else{
                                console.log("ERROR: " + data);
                            }
                        });
                    //-----------------------------------------
                    //-----------------------------------------

                    //Redirect User to Home
                    // this.props.history.push('/');
                })
                .catch((error) => {
                    //Failure
                    //TODO: Cleanup Firebase Error Messages
                    this.setState({ submitError: error.message });
                });
        }
        else {
            //Display field validation errors
            this.setState({ errors: errors });
        }

        event.preventDefault();
    }

    onChange = event => {
        this.setState({ [event.target.name]: event.target.value });
    };

    render() {
        const {
            email,
            passwordOne,
            passwordTwo,
            errors,
            submitError,
        } = this.state;

        const isInvalid =
            // passwordOne !== passwordTwo ||
            passwordOne === '' ||
            passwordTwo === '' ||
            email === '';

        const { classes } = this.props;

        return (
            <form className={classes.form} onSubmit={this.onSubmit}>
                <Typography component="h6" className={classes.errorMsg} >
                    {submitError}
                </Typography>
                <TextField
                    variant="outlined"
                    margin="normal"
                    required
                    fullWidth
                    name="email"
                    value={email}
                    onChange={this.onChange}
                    type="text"
                    label="Email Address"
                    error={errors["email"] !== ""}
                    helperText={errors["email"]}
                />
                <TextField
                    variant="outlined"
                    margin="normal"
                    required
                    fullWidth
                    name="passwordOne"
                    value={passwordOne}
                    onChange={this.onChange}
                    type="password"
                    label="Password"
                    error={errors["passwordOne"] !== ""}
                    helperText={errors["passwordOne"]}
                />
                <TextField
                    variant="outlined"
                    margin="normal"
                    required
                    fullWidth
                    name="passwordTwo"
                    value={passwordTwo}
                    onChange={this.onChange}
                    type="password"
                    label="Confirm Password"
                    error={errors["passwordTwo"] !== ""}
                    helperText={errors["passwordTwo"]}
                />
                <Button className={classes.submit}
                    type="submit"
                    fullWidth
                    variant="contained"
                    color="primary"
                    disabled={isInvalid}
                >
                    Sign Up
                        </Button>
            </form>
        );
    }
}

const SignInForgotLink = ({ classes }) => (
    <Grid container>
        <Grid item xs>
            <Link to='/forgot' className={classes.links}>
                <MUILink variant="body2" component={'span'}>
                    Forgot password?
                </MUILink>
            </Link>
        </Grid>
        <Grid item>
            <Link to='/signin' className={classes.links}>
                <MUILink variant="body2" component={'span'}>
                    Already have an account? Sign In
                </MUILink>
            </Link>
        </Grid>
    </Grid>
);

const SignUpForm = withRouter(SignUpFormBase);

export default withStyles(styles)(SignUp);

export { SignUpForm, SignInForgotLink };