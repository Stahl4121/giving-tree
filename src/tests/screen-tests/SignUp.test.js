import React from 'react';
import { shallow, mount } from 'enzyme';
import { Link, MemoryRouter } from 'react-router';
import { StaticRouter } from 'react-router-dom';
import firebase from '../../firebase.js';

import SignUp from '../../screens/SignUp.js';

describe('SignUp with no props', () => {
  const context = {};
  const container = mount(
    <StaticRouter location="someLocation" context={context}>
        <SignUp />
    </StaticRouter>);

  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });
 
  it('should have an email field', () => {
    expect(container.find('#email').at(0).length).toEqual(1);
  });

  it('should have proper props for email field', () => {
    expect(container.find('#email').at(0).props()).toEqual({
      variant: "outlined",
      margin: "normal",
      required: true,
      fullWidth: true,
      name: "email",
      onChange: expect.any(Function),
      type: "text",
      label: "Email Address",
      autoFocus: true,
      error: false,
      helperText: "",
      id: "email",
      value: "", 
    });
  });

  it('should have a passwordOne field', () => {
    expect(container.find('#passwordOne').at(0).length).toEqual(1);
  });

  it('should have proper props for passwordOne field', () => {
    expect(container.find('#passwordOne').at(0).props()).toEqual({
      error: false,
      variant: "outlined",
      margin: "normal",
      required: true,
      fullWidth: true,
      name: "passwordOne",
      onChange: expect.any(Function),
      type: "password",
      label: "Password",
      id: "passwordOne",
      helperText: "",
      value: "", 
    });
  });

  it('should have a passwordConfimrm field', () => {
    expect(container.find('#passwordTwo').at(0).length).toEqual(1);
  });

  it('should have proper props for passwordTwo field', () => {
    expect(container.find('#passwordTwo').at(0).props()).toEqual({
      error: false,
      helperText: "",
      variant: "outlined",
      margin: "normal",
      required: true,
      fullWidth: true,
      name: "passwordTwo",
      onChange: expect.any(Function),
      type: "password",
      label: "Confirm Password",
      id: "passwordTwo",
      value: "", 
    });
  });

  it('should have a submit button', () => {
    expect(container.find('#submit-button').at(0).length).toEqual(1);
  });

  it('should have proper props for the submit button', () => {
    expect(container.find('#submit-button').at(0).props()).toEqual({
      variant: "contained",
      children: "Sign Up",
      className: "SignUp-submit-4",
      color: "primary",
      disabled: true,
      fullWidth: true,
      type: "submit",
      id: "submit-button",
    });
  });

});

describe('SignUp with props', () => {
  let context, container;

  beforeEach(() => {
    context = {};
    container = mount(
      <StaticRouter location="someLocation" context={context}>
        <SignUp />
      </StaticRouter>);
  });

  it('should set the email value on change event', () => {
    container.find('#email').at(0).props().onChange({
      target: {
        name: 'email',
        value: 'acesmndr@gmail.com',
      },
    });
    expect(container.find('#email').at(0).length).toEqual(1);
    expect(container.find('SignUpFormBase').state()).toEqual({
      "email": 'acesmndr@gmail.com',
      "errors": {
        "email": "",
        "passwordOne": "",
        "passwordTwo": "",
        "submit": "",
      },
      "isValid": false,
      "passwordOne": '',
      "passwordTwo": ''
    });
  });

  it('should set the password value on change event', () => {
    container.find('#passwordOne').at(0).props().onChange({
      target: {
        name: 'passwordOne',
        value: 'notapassword',
      },
    });
    expect(container.find('#passwordOne').at(0).length).toEqual(1);
    expect(container.find('SignUpFormBase').state()).toEqual({
      "email": "",
      "errors": {
        "email": "",
        "passwordOne": "*Password fails the following requirements: a number, an uppercase letter, a special character (e.g. !@#$%^&*)",
        "passwordTwo": "",
        "submit": "",
      },
      "isValid": false,
      "passwordOne": 'notapassword',
      "passwordTwo": ''
    });

  });

  it('should set the password check value on change event', () => {
    container.find('#passwordTwo').at(0).props().onChange({
      target: {
        name: 'passwordTwo',
        value: 'notapassword',
      },
    });
    expect(container.find('#passwordTwo').at(0).length).toEqual(1);
    expect(container.find('SignUpFormBase').state()).toEqual({
      "email": "",
      "errors": {
        "email": "",
        "passwordOne": "",
        "passwordTwo": "*Passwords do not match.",
        "submit": "",
      },
      "isValid": false,
      "passwordOne": '',
      "passwordTwo": 'notapassword'
    });

  });

    it('should have proper props for the submit button', () => {
      expect(container.find('#submit-button').at(0).props()).toEqual({
        variant: "contained",
        children: "Sign Up",
        className: "SignUp-submit-4",
        color: "primary",
        disabled: true,
        fullWidth: true,
        type: "submit",
        id: "submit-button",
      });
    });
  
  });

  describe('SignUp with props', () => {
    let context, container;
  
    beforeEach(() => {
      context = {};
      container = mount(
        <StaticRouter location="someLocation" context={context}>
          <SignUp />
        </StaticRouter>);
        firebase.auth = jest.fn().mockReturnValue({
          createUserWithEmailAndPassword: function(email, passwordOne) { return true; }
          });
    });
  
    it('should allow submit when given correct input', () => {
      container.find('#email').at(0).props().onChange({
        target: {
          name: 'email',
          value: 'acesmndr@gmail.com',
        },
      });
      expect(container.find('#email').at(0).length).toEqual(1);
      
      container.find('#passwordOne').at(0).props().onChange({
        target: {
          name: 'passwordOne',
          value: '1q!Q1q!q',
        },
      });
      expect(container.find('#passwordOne').at(0).length).toEqual(1);
      
      container.find('#passwordTwo').at(0).props().onChange({
        target: {
          name: 'passwordTwo',
          value: '1q!Q1q!q',
        },
      });
      expect(container.find('#passwordTwo').at(0).length).toEqual(1);
      
      expect(container.find('SignUpFormBase').state()).toEqual({
        "email": "acesmndr@gmail.com",
        "errors": {
          "email": "",
          "passwordOne": "",
          "passwordTwo": "",
          "submit": "",
        },
        "isValid": true,
        "passwordOne": '1q!Q1q!q',
        "passwordTwo": '1q!Q1q!q'
      });
      
      expect(container.find('#submit-button').at(0).props()).toEqual({
        variant: "contained",
        children: "Sign Up",
        className: "SignUp-submit-4",
        color: "primary",
        disabled: true,
        fullWidth: true,
        type: "submit",
        id: "submit-button",
      });

      //let result = container.onSubmit();
      //expect(result).toBe(true);
    });
  });