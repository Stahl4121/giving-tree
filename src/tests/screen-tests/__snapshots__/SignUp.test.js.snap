// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp with no props should match the snapshot 1`] = `"<main class=\\"MuiContainer-root MuiContainer-maxWidthXs\\"><div class=\\"SignUp-root-1\\"><div class=\\"MuiAvatar-root MuiAvatar-circle SignUp-avatar-2 MuiAvatar-colorDefault\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z\\"></path></svg></div><h1 class=\\"MuiTypography-root MuiTypography-h5\\">Sign Up</h1><form class=\\"SignUp-form-3\\" novalidate=\\"\\"><h6 class=\\"MuiTypography-root SignUp-errorMsg-6 MuiTypography-body1\\"></h6><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"false\\" for=\\"email\\" id=\\"email-label\\">Email Address<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"email\\" name=\\"email\\" required=\\"\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-119 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-121\\"><span>Email Address&nbsp;*</span></legend></fieldset></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"false\\" for=\\"passwordOne\\" id=\\"passwordOne-label\\">Password<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"passwordOne\\" name=\\"passwordOne\\" required=\\"\\" type=\\"password\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-119 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-121\\"><span>Password&nbsp;*</span></legend></fieldset></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"false\\" for=\\"passwordTwo\\" id=\\"passwordTwo-label\\">Confirm Password<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"passwordTwo\\" name=\\"passwordTwo\\" required=\\"\\" type=\\"password\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-119 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-121\\"><span>Confirm Password&nbsp;*</span></legend></fieldset></div></div><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained SignUp-submit-4 MuiButton-containedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled\\" tabindex=\\"-1\\" type=\\"submit\\" disabled=\\"\\" id=\\"submit-button\\"><span class=\\"MuiButton-label\\">Sign Up</span></button></form><div class=\\"MuiGrid-root MuiGrid-container\\"><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorPrimary\\" href=\\"/forgot\\">Forgot password?</a></div><div class=\\"MuiGrid-root MuiGrid-item\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorPrimary\\" href=\\"/signin\\">Already have an account? Sign In</a></div></div></div></main>"`;
